src/osdep/osdep.c

```diff
@@ -1018,3 +1018,80 @@ get_system_str(int xwin)
 {
 	return xwin ? SYSTEM_STR "-xwin" : SYSTEM_STR;
 }
+
+#if _DEFAULT_SOURCE || _SVID_SOURCE || _BSD_SOURCE
+
+/* tempnam() replacement without races */
+
+int isdirectory(char *path) {
+	struct stat ss;
+	if (path == NULL)
+		return 0;
+	if (-1 == stat(path, &ss))
+		return 0;
+	return S_ISDIR(ss.st_mode);
+}
+
+char *tempname(char *dir, char *pfx, char *suff) {
+	struct string path;
+	int fd;
+
+	if (isdirectory(getenv("TMPDIR")))
+		dir = getenv("TMPDIR");
+	else if (dir != NULL)
+		dir = dir;
+	else if (isdirectory(P_tmpdir))
+		dir = P_tmpdir;
+	else if (isdirectory("/tmp"))
+		dir = "/tmp";
+	else {
+		errno = ENOTDIR;
+		return NULL;
+	}
+
+	if (!init_string(&path)) {
+		errno = ENOMEM;
+		return NULL;
+	}
+	add_to_string(&path, dir);
+	add_to_string(&path, "/");
+	add_to_string(&path, "/");
+	add_to_string(&path, pfx);
+	add_to_string(&path, "XXXXXX");
+	if (suff)
+		add_shell_safe_to_string(&path, suff, strlen(suff));
+
+	fd = mkstemps(path.source, suff ? strlen(suff) : 0);
+	if (fd == -1) {
+		done_string(&path);
+		errno = ENOENT;
+		return NULL;
+	}
+	close(fd);
```

Comment:
It is not so easy. AFAIK, the problem with tmpnam is such, that between obtaining name, and opening it with open, something or someone can open this file first.
Here when you close fd, there is the same problem. Somebody can open this file first.
Solution with mkstemps must work with file descriptors. So you should return fd of open file, and change code in other places where tempnam was used to accept file descriptor instead of name.

I do not think tempnam is a big problem, especially if you set TMPDIR to some place
in your HOME directory.

Reply:
By this token, someone else may open the file between file creation and writing. I believe the problem is rather than someone else may create the file in between the filename is returned by tempnam() and the file is created. This way, they can create the file with their username and permissions to make its content readable.

This is not a problem with this commit because the file is created at the same time its file name is returned. Rather, the problem is that someone else may delete and recreate it before elinks opens it again. Yet, this may also happen when passing the file name to an external application.

That made me think about what I was actually trying to solve. It wasn't insecurity, but the possibility that the temp file is accidentally created in the same way: "The tempnam() function returns a pointer to a string that is a valid filename, and such that a file with this name did not exist when tempnam() checked.", from tempnam(1). This means that between tempnam() and open(), another application (another instance of elinks itself, given the file pattern) may obtain the same filename from tempnam(). By creating the file at the same time as its name is obtained, such clash is impossible.

I changed the title of the commit to convey that: it creates the file at the same time its name is chosen.